[docs] Add functions-develop file for Pulsar Functions(new)

Jennifer88huang opened the pull request:
Fixes #4554 

### Motivation
Add a new develop tutorial for new Pulsar Functions structure.

### Modifications
1. Readjust the structure of functions-develop.
2. Adopt the language-specific code tabs, and leave [placeholder] for parts of "Go/Python/Java". If there are, please provide relevant examples. If no, please comment, and I'll remove those tabs.
3. Ignore links whether they're right or wrong, regard them as placeholder, and I'll check and refine throughout when other parts are all finished.

Jennifer88huang commented:
@wolfstudy Could you please help review? Thank you.

jerrypeng commented on site2/docs/functions-develop.md:
````diff
@@ -0,0 +1,570 @@
+    return "{}!".format(input)
+```
+
+However, the function uses Pulsar Functions [SDK for Python](#python-sdk-functions):
````
This sentence is a bit awkward

Jennifer88huang replied:
I've refined it.

jerrypeng commented on site2/docs/functions-develop.md:
```diff
@@ -0,0 +1,570 @@
+
+This tutorial walks you through how to develop Pulsar Functions.
+
+## Available APIs
```
For this section, lets have examples both language native and SDK for all languages. Currently there is only python and GO

jerrypeng commented on site2/docs/functions-develop.md:
```diff
@@ -0,0 +1,570 @@
+Pulsar has a built in [Schema Registry](concepts-schema-registry) and comes bundled with a variety of popular schema types(avro, json and protobuf). Pulsar Functions can leverage existing schema information from input topics and derive the input type. The schema registry applies for output topic as well.
+
+## SerDe
+SerDe stands for **Ser**ialization and **De**serialization. Pulsar Functions use SerDe when publishing data to and consuming data from Pulsar topics. How SerDe works by default depends on the language you use for a particular function.
```
"Pulsar Functions" here refers to the framework.  Should it be followed by a singular verb i.e. "uses"?

jerrypeng commented on site2/docs/functions-develop.md:
````diff
@@ -0,0 +1,570 @@
+}
+```
+
+<!--Go-->
````
Should we provide some context that this is the GO API

jerrypeng commented on site2/docs/functions-develop.md:
```diff
@@ -0,0 +1,570 @@
+Language-native interface | No Pulsar-specific libraries or special dependencies required (only core libraries from Java/Python/Go). | Functions that do not require access to the function [context](#context).
+Pulsar Function SDK for Java/Python/Go | Pulsar-specific libraries that provide a range of functionality not provided by "native" interfaces. | Functions that require access to the function [context](#context).
+
+<!--DOCUSAURUS_CODE_TABS-->
```
Can we have a separate section for examples using the SDK API.  I don't think its the best to coalesce language native and SDK examples together

jerrypeng replied:
If we can also provide links to the complete code, I think that would also be help to the reader

Jennifer88huang replied:
Sure, separate them.

jerrypeng commented on site2/docs/functions-develop.md:
```diff
@@ -0,0 +1,570 @@
+Pulsar Function SDK for Java/Python/Go | Pulsar-specific libraries that provide a range of functionality not provided by "native" interfaces. | Functions that require access to the function [context](#context).
+
+<!--DOCUSAURUS_CODE_TABS-->
+<!--Java-->
```
Java native:

```
public class JavaNativeExclamationFunction implements Function<String, String> {
    @Override
    public String apply(String input) {
        return String.format("%s!", input);
    }
}
```
Link to complete code: 
https://github.com/apache/pulsar/blob/master/pulsar-functions/java-examples/src/main/java/org/apache/pulsar/functions/api/examples/ExclamationFunction.java
https://github.com/apache/pulsar/blob/master/pulsar-functions/java-examples/src/main/java/org/apache/pulsar/functions/api/examples/JavaNativeExclamationFunction.java


SDK:

```
public class ExclamationFunction implements Function<String, String> {
    @Override
    public String process(String input, Context context) {
        return String.format("%s!", input);
    }
}
```

Link to complete code:

jerrypeng commented on site2/docs/functions-develop.md:
```diff
@@ -0,0 +1,570 @@
+<!--DOCUSAURUS_CODE_TABS-->
+<!--Java-->
+<!--Python-->
+In Python, the language-native function, which adds an exclamation point to all incoming strings and publishes the resulting string to a topic, has no external dependencies.
```
We can probably just mention this before the the code examples for all languages.  There is no point to mention this again and again for each language

Jennifer88huang replied:
Sure. Good idea.

jerrypeng commented on site2/docs/functions-develop.md:
````diff
@@ -0,0 +1,570 @@
+The following example uses Pulsar Functions SDK.
+
+```
+import (
````
@wolfstudy you provide a language native example for GO.  Preferably one that matches the exclamation examples we have for the other languages.

wolfstudy replied:
Sorry, Go Function does not support schema-related functions, so now Go Function's input and output can only be `[]byte`. 

Here, this [code example](https://github.com/apache/pulsar/blob/master/pulsar-function-go/examples/inputFunc.go#L20-L36) might be more appropriate:
```
package main

import (
	"context"
	"fmt"

	"github.com/apache/pulsar/pulsar-function-go/pf"
)

func HandleRequest(ctx context.Context, in []byte) error{
	fmt.Println(string(in) + "!")
	return nil
}

func main() {
	pf.Start(HandleRequest)
}
```

Jennifer88huang commented:
@jerrypeng thank you very much for all your comments, I've updated according to your advice.
I've added the complete code link for Python native and SDK examples, and found the example content in this file is different from that in the links. I'm not sure I've used the correct link. Please help check line 36 and 68.
If the links are wrong, feel free to correct.
@wolfstudy Could you please provide Go examples and relative links? Thank you.

Jennifer88huang commented:
@jerrypeng @wolfstudy Thank you very much for your comments, I've updated the content accordingly.
1. Update Python SDK example, to make it consistent with what provided in the code link.
2. Update Go SDK example as provided by @wolfstudy

Jennifer88huang commented:
@jerrypeng @sijie Is there any other issue in this file?

sijie commented on site2/docs/functions-develop.md:
````diff
@@ -0,0 +1,594 @@
+          tweet_components = str(input_bytes).split('|')
+          return Tweet(tweet_components[0], tweet_componentsp[1])
+  ```
+<!--Go-->
````
@Jennifer88huang Can you create a github issue for adding `Go` serde option?

Also is it worth adding a statement here indicating the documentation for Go serde is not available yet.

Jennifer88huang replied:
I'll create an issue for this. And my suggestion is just remove the "Go" tab if it's not available yet. When there is no such tab, readers know it's not available. When it's available, we can add it immediately.

Jennifer88huang replied:
For issue, see https://github.com/apache/pulsar/issues/4752

sijie commented on site2/docs/functions-develop.md:
```diff
@@ -0,0 +1,594 @@
+In both languages, however, you can write custom SerDe logic for more complex, application-specific types.
+
+## Context
+Both the [Java](#java-sdk-functions) and [Python](#python-sdk-functions) SDKs provide access to a **context object** that can be used by a function. This context object provides a wide variety of information and functionality to the function.
```
Go also supports context, right?

wolfstudy replied:
right, Go Function supports context logic.

sijie commented on site2/docs/functions-develop.md:
````diff
@@ -0,0 +1,594 @@
+        else:
+            logger.info("The word of the day is {0}".format(wotd))
+```
+
````
Although Go doesn't support UserConfig yet, but it would be great adding a tab here indicating "the feature is not available in Go".

sijie commented on site2/docs/functions-develop.md:
````diff
@@ -0,0 +1,594 @@
+```
+
+All logs produced by `LoggingFunction` above can be accessed via the `logging-function-logs` topic.
+
````
Add Go as well?

sijie commented on site2/docs/functions-develop.md:
````diff
@@ -0,0 +1,594 @@
+            context.record_metric('elevens-count', 1)
+```
+
+<!--Go-->
````
Add a statement indicating "the feature is not available in Go yet".

Jennifer88huang replied:
I've removed the tab for "Go".

sijie commented on site2/docs/functions-develop.md:
```diff
@@ -0,0 +1,594 @@
+Since Pulsar 2.1.0 release, Pulsar integrates with Apache BookKeeper [table service](https://docs.google.com/document/d/155xAwWv5IdOitHh1NVMEwCMGgB28M3FyMiQSxEpjE-Y/edit#heading=h.56rbh52koe3f) to store the `State` for functions. For example, a `WordCount` function can store its `counters` state into BookKeeper table service via Pulsar Functions State API.
+
+## Secrets/Encryption
+[Need information here]
```
- Please create a github issue for adding documentation for secrets and encryption. @jerrypeng can help with providing the content.

- Add a statement here indicating the section is under construction ?

Jennifer88huang replied:
Maybe we can use "Encryption" here?
For issues, see https://github.com/apache/pulsar/issues/4753

Jennifer88huang commented:
@sijie Since some features are not supported in Go at the moment, I just remove the tab for "Go".
When the feature is available, we can add the content immediately.
If we write "xxx is not available/supported..." too many times, it may lead to negative effect for users.
When we do not add the tab, it is not supported in Go by default.
How do you think about this?

sijie commented:
@Jennifer88huang sgtm

Jennifer88huang commented:
I've created issue to add "Encryption" part in https://github.com/apache/pulsar/issues/4753, could we merge this PR if there are no other issues?

Jennifer88huang commented:
run java8 tests

sijie commented:
retest this please

Jennifer88huang commented:
run integration tests
run java8 tests
